### PR TITLE
Change DBHandler to split transport and query running

### PR DESCRIPTION
**Pull request by igalink**
Draft to split transport and query running in DBHandler

### CODE REVIEW

**Review by ienkovich — CHANGES_REQUESTED** [review]
I cannot build this branch with multiple errors in DBHandler.

**Comment by igalink** [comment]
> I cannot build this branch with multiple errors in DBHandler.

Fixed

**Review by ienkovich — COMMENTED** [review]
I think the patch looks good enough. Let's commit as long as unit tests pass.

**Comment by anmyachev** [comment]
@ienkovich can we merge this PR?

**Comment by ienkovich** [comment]
As I said, this PR is ok to merge if it passes all tests.

**Comment by anmyachev** [comment]
@ienkovich where can I see the status of the tests?

**Comment by ienkovich** [comment]
That's what Galina should let us know.

**Comment by anmyachev** [comment]
Thanks @ienkovich. 

@igalink  can you share status with us?
> @ienkovich where can I see the status of the tests?

**Comment by igalink** [comment]
Tests has not passed yet. I'm fixing issue with read_csv python test.

File: Embedded/DBEngine.cpp [hunk 1]
```diff
@@ -341,49 +365,39 @@ class DBEngineImpl : public DBEngine {
 
   bool setDatabase(std::string& db_name) {
     auto& sys_cat = Catalog_Namespace::SysCatalog::instance();
-    auto catalog = sys_cat.switchDatabase(db_name, user_.userName);
-    updateSession(catalog);
-    sys_cat.getMetadataForDB(db_name, database_);
+    auto user = db_handler_->get_session_copy(session_id_).get_currentUser();
+    auto catalog = sys_cat.switchDatabase(db_name, user.userName);
```
Comment (fexolm) [inline Embedded/DBEngine.cpp:369]: `catalog` seems to be unused, also, it looks like we could avoid use `auto &user` here

File: Embedded/DBEngine.cpp [hunk 2]
```diff
@@ -452,11 +466,9 @@ class DBEngineImpl : public DBEngine {
 
  private:
   std::string base_path_;
+  std::string session_id_;
   std::shared_ptr<ForeignStorageInterface> fsi_;
-  std::shared_ptr<Data_Namespace::DataMgr> data_mgr_;
-  std::shared_ptr<Calcite> calcite_;
-  Catalog_Namespace::DBMetadata database_;
-  Catalog_Namespace::UserMetadata user_;
+  mapd::shared_ptr<DBHandler> db_handler_;
```
Comment (fexolm) [inline Embedded/DBEngine.cpp:471]: is there any reason to use mapd::shared_ptr only in that place?

File: Embedded/DBEngine.cpp [hunk 3]
```diff
@@ -469,50 +481,10 @@ namespace {
...
-  g_enable_columnar_output = true;
-  for (const auto& [key, value] : parameters) {
-    if (key == "path") {
-      path = value;
-    } else if (key == "port") {
-      port = std::stoi(value);
-    } else if (key == "enable_columnar_output") {
-      g_enable_columnar_output = std::stoi(value);
-    } else if (key == "enable_union") {
-      g_enable_union = std::stoi(value);
```
Comment (fexolm) [inline Embedded/DBEngine.cpp:499]: where all those options are configured now?

File: QueryEngine/Descriptors/RelAlgExecutionDescriptor.cpp [hunk 4]
```diff
@@ -21,28 +21,64 @@
 #include "QueryEngine/GroupByAndAggregate.h"
 #include "QueryEngine/RelAlgDagBuilder.h"
 
+ExecutionResult::ExecutionResult()
+    : results_(std::make_shared<ResultSet>(std::vector<TargetInfo>{},
```
Comment (fexolm) [inline QueryEngine/Descriptors/RelAlgExecutionDescriptor.cpp:25]: do we need an empty result? why?

File: QueryEngine/Descriptors/RelAlgExecutionDescriptor.h [hunk 5]
```diff
@@ -72,13 +74,30 @@ class ExecutionResult {
     results_[0]->setQueueTime(queue_time_ms);
   }
 
+  enum RType { QueryResult, SimpleResult, Explaination, CalciteDdl };
+
+  std::string getExplanation();
+  void updateResultSet(const std::string& query_ra, RType type, bool success = true);
+  RType getResultType() const { return type_; }
```
Comment (fexolm) [inline QueryEngine/Descriptors/RelAlgExecutionDescriptor.h:81]: why do we need that mutability support now?

File: Embedded/DBEngine.cpp [hunk 6]
```diff
@@ -102,73 +101,132 @@ class DBEngineImpl : public DBEngine {
...
-                                                   calcite_,
-                                                   false);
-    sys_cat.getMetadataForUser(OMNISCI_ROOT_USER, user_);
-    auto session = std::make_unique<Catalog_Namespace::SessionInfo>(
-        catalog, user_, ExecutorDeviceType::CPU, "");
-    QR::init(session);
-    base_path_ = db_path;
+    try {
+      db_handler_ = mapd::make_shared<DBHandler>(
+              prog_config_opts.db_leaves,
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply Embedded/DBEngine.cpp:153]: That would be a general improvement for DBHandler. Nice to have but not related to the implemented feature.

File: Embedded/DBEngine.cpp (hunk 2 above)
Comment (ienkovich) [reply Embedded/DBEngine.cpp:471]: This is to support older Thrift versions.

File: Embedded/DBEngine.cpp (hunk 3 above)
Comment (ienkovich) [reply Embedded/DBEngine.cpp:499]: We now use CommandLineOptions in init method to configure all parameters.

File: QueryEngine/Descriptors/RelAlgExecutionDescriptor.cpp (hunk 4 above)
Comment (ienkovich) [reply QueryEngine/Descriptors/RelAlgExecutionDescriptor.cpp:25]: ExecutionResult is supposed to have a result set. There are many places in the code where ExecutionResult is constructed with the empty result set, we just add a special constructor for convenience.

File: QueryEngine/Descriptors/RelAlgExecutionDescriptor.h (hunk 5 above)
Comment (ienkovich) [reply QueryEngine/Descriptors/RelAlgExecutionDescriptor.h:81]: We now use ExecutionResult as an output parameter and have to modify it.

**Review by ienkovich — APPROVED** [review]
LGTM!

**Review by fexolm — APPROVED** [review]

File: Embedded/DBEngine.cpp (hunk 1 above)
Comment (igalink) [reply Embedded/DBEngine.cpp]: Fixed

File: Embedded/Python/dbe.pyx [hunk 7]
```diff
@@ -163,9 +163,9 @@ cdef class PyDbEngine:
     cdef map[string, string] c_parameters
 
     def __cinit__(self, **kwargs):
-        for key, value in kwargs.items():
-            self.c_parameters[key] = str(value)
-        self.c_dbe = DBEngine.create(self.c_parameters)
+        cmd_str = "".join(' --%s %r' % x for x in kwargs.iteritems())
+        cmd_str = cmd_str.replace("_", "-")
```
Comment (ienkovich) [inline Embedded/Python/dbe.pyx:167]: This replacement should affect keys only. E.g. for `--data=data_tmp` you would get `--data=data-tmp` which is not the desired behavior.

Comment (igalink) [reply Embedded/Python/dbe.pyx]: Fixed

**Review by ienkovich — APPROVED** [review]

**Pull request merged**